src/aof.c

```diff
@@ -928,7 +1317,98 @@ int loadAppendOnlyFile(char *filename) {
 cleanup:
     if (fakeClient) freeClient(fakeClient);
     fclose(fp);
-    stopLoading(ret == AOF_OK);
+    return ret;
+}
+
+/* Load the AOF files according the aofManifest pointed by am. */
+int loadAppendOnlyFiles(aofManifest *am) {
+    serverAssert(am != NULL);
+    int ret = C_OK;
+    long long start;
+    off_t base_aof_size = 0, incr_aof_size = 0, size;
+    char *aof_name;
+
+    /* If there is no information about BASE AOF and INCR AOF in aofManifest, then 
+     * there is only one possibility: the aof manifest file does not exist, we may 
+     * be starting from an old redis version. so we must fall back to the previous 
+     * loading mode.
+     */
+    if (am->base_aof_info == NULL && listLength(am->incr_aof_list) == 0) {
+        struct redis_stat sb;
+        /* If the `server.aof_filename` file does not exist, we will directly return 
+         * AOF_NOT_EXIST, and redis will ignore this error. */
+        if (redis_stat(server.aof_filename, &sb) != 0) {
+            return AOF_NOT_EXIST;
+        }
+
+        /* If the server.aof_filename file exists, we manually construct a BASE 
+         * type aofInfo and add it to aofManifest. In this way, we can reuse the 
+         * following code to load this AOF file. */
+        aofInfo *ai = aofInfoCreate();
+        ai->file_name = sdsnew(server.aof_filename);
+        ai->file_seq = 1;
+        ai->file_type = AOF_FILE_TYPE_BASE;
+        am->base_aof_info = ai;
+        am->curr_base_aof_seq = 1;
+        am->dirty = 1;
+
+        int ret = persistAofManifest(am);
+        if (ret == C_ERR) {
+            return AOF_FAILED;
+        }
+    }
+
+    /* Load BASE AOF. */
+    if (am->base_aof_info) {
+        serverAssert(am->base_aof_info->file_type == AOF_FILE_TYPE_BASE);
+        aof_name = (char*)am->base_aof_info->file_name;
+        size = getAppendOnlyFileSize(aof_name);
+        start = ustime();
+        startLoading(size, RDBFLAGS_AOF_PREAMBLE, 0);
```

Comment:
when loading an RDB, i think we may want to set `rdbFileBeingLoaded` (which `startLoadingFile` sets) in some way,
this controls behavior when corruption is detected.
and it would also be a good idea to pass the file size.
on the other hand, i don't think we wanna do a separate startLoading / stopLoading per file, and instead use just one pair like before.
so we need to find a way to determine the file size ahead of time, and update the file name when needed.
let's look into it and see what we can come up with.

Reply:
Modified